### PR TITLE
Preventing early materialization of query on load

### PR DESCRIPTION
**Pull request by isha97**

### CODE REVIEW

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java [hunk 1]
```diff
@@ -630,6 +643,23 @@ public TableInfo materializeViewToTable(
...
 
+  public Schema getTableSchema(String querySql, Map<String, String> additionalQueryJobLabels) {
+    JobInfo jobInfo =
+        JobInfo.of(
+            jobConfigurationFactory
+                .createQueryJobConfigurationBuilder(querySql, additionalQueryJobLabels)
+                .setDryRun(true)
+                .build());
+
+    log.info("running query {}", querySql);
```
Comment (vishalkarve15) [inline bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java:654]: Can log "dry running query" to differentiate between actual query run and this dry run.

File: spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/integration/ReadFromQueryIntegrationTestBase.java [hunk 2]
```diff
@@ -118,6 +131,26 @@ public void testQueryOption() {
...
 
+    StructType expectedSchema =
+        DataTypes.createStructType(
+            ImmutableList.of(
+                DataTypes.createStructField("corpus", DataTypes.StringType, true),
+                DataTypes.createStructField("word_count", DataTypes.LongType, true)));
+
+    assertThat(df.schema()).isEqualTo(expectedSchema);
+
+    if (isSpark3AndAbove) {
```
Comment (vishalkarve15) [inline spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/integration/ReadFromQueryIntegrationTestBase.java:142]: It's actually for dsv2 right? Doesn't need to be Spark 3 and above. Can rename to `isDataSourceV2`.

**Comment by vishalkarve15** [comment]
Please add to CHANGES.md

File: spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/integration/ReadFromQueryIntegrationTestBase.java (hunk 2 above)
Comment (davidrabinowitz) [reply spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/integration/ReadFromQueryIntegrationTestBase.java:142]: it's DSv2 on Spark 3.x ...

File: spark-bigquery-dsv2/spark-3.1-bigquery-lib/src/main/java/com/google/cloud/spark/bigquery/v2/Spark31BigQueryTable.java [hunk 3]
```diff
@@ -67,7 +65,7 @@ protected BigQueryDataSourceReaderContext createBigQueryDataSourceReaderContext(
             ImmutableMap.of(),
             injector.getInstance(DataSourceVersion.class),
             injector.getInstance(SparkSession.class),
-            Optional.of(schema), /*tableIsMandatory*/
+            Optional.of(schemaSupplier.get()), /*tableIsMandatory*/
```
Comment (davidrabinowitz) [inline spark-bigquery-dsv2/spark-3.1-bigquery-lib/src/main/java/com/google/cloud/spark/bigquery/v2/Spark31BigQueryTable.java:68]: Should be `Optional.ofNullable()`, as you may return `null`

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: CHANGES.md [hunk 4]
```diff
@@ -5,6 +5,7 @@
 ## 0.35.1 - 2023-12-28
 
 * PR #1153: allow writing spark string to BQ datetime
+* PR #1155: allow lazy materialization of query on load
```
Comment (davidrabinowitz) [inline CHANGES.md:8]: Should be under the `Next` section

**Review by davidrabinowitz — APPROVED** [review]

File: CHANGES.md (hunk 4 above)
Comment (isha97) [reply CHANGES.md]: Yes. fixed.

**Review by vishalkarve15 — APPROVED** [review]

**Pull request merged**